### PR TITLE
#60 typeorm db migrations in api project

### LINKED ISSUE #60: TypeORM & DB migrations in api project
TpeORM is a conservative choice, and migrations are a safe option for large-scale projects.
This should have basic implementation, and make commands prepared for migrations creation, execution and other options that might be useful and also documentation.

### PR DESCRIPTION
**Pull request by vygandas**
### What does it do?

Added TypeORM for working with databases and entities and also added TypeORM Migrations commands. This of course has to be extended to cover running migrations, not only creating.

### Why is it needed?

It's needed because we need to store data 😁 

### How to test it?

Either works or not.

Inspiration for implementation https://github.com/typeorm/typeorm/issues/9122

### CODE REVIEW

**Comment by llong2195** [comment]
I'm looking forward to this project, thank you for creating it.
Ormconfig doesn't work correctly with webpack.

> `entities: [join(__dirname, '**', '*.entity.{ts,js}')],`

> `migrations: [join(__dirname, '**', '*.migration.{ts,js}')],`

Expected Behavior:

> All classes found using the provided glob pattern...

Actual Behavior:

> No classes were found using the provided glob pattern...

Resolution:
1. `entities: getMetadataArgsStorage().tables.map((tbl) => tbl.target),`
2. Instead of building with webpack, build using the NestJS CLI."

File: apps/api/src/config/typeorm.config.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  type: 'postgres',
+  host: `${String(process.env.DATABASE_HOST)}`,
+  port: `${String(process.env.DATABASE_PORT)}`,
+  username: `${String(process.env.DATABASE_USERNAME)}`,
+  password: `${String(process.env.DATABASE_PASSWORD)}`,
+  database: `${String(process.env.DATABASE_NAME)}`,
+  // entities: ['dist/**/*.entity{.ts,.js}'],
+  // migrations: ['dist/migrations/*{.ts,.js}'],
+
+  entities: [join(__dirname, '**', '*.entity.{ts,js}')],
```
Comment (mantaskaveckas) [inline apps/api/src/config/typeorm.config.ts:17]: Don't need to pass entities here since `autoLoadEntities` is `true`.

**Comment by vygandas** [comment]
@llong2195 could you tell me how exactly you found this issue? Steps to reproduce would be very helpful to me :) For example, migrations work fine, which means it is being captured properly.

**Comment by vygandas** [comment]
@llong2195 ah, it might be my bad because migration was not captured properly, it was a different SQL log haha.

**Comment by llong2195** [comment]
@vygandas i created mini repo: [issue-webpack-typeorm](https://github.com/llong2195/issue-webpack-typeorm)

**Comment by vygandas** [comment]
@llong2195 , I've fixed migrations, now it's working fine. What's left is entities. At the moment, it's not being captured. Getting this
```
[Nest] 53159  - 11/24/2023, 9:40:26 AM   ERROR [ExceptionsHandler] No metadata for "User" was found.
```
This needs some thinking.

I've also added you contributors, so if you see how this can be fixed, you can push to this branch. Also reviewing your repository that you've sent.

PS. Let's keep option `synchronize: false,`, since we are using migrations.

**Comment by llong2195** [comment]
> @llong2195 , I've fixed migrations, now it's working fine. What's left is entities. At the moment, it's not being captured. Getting this
> 
> ```
> [Nest] 53159  - 11/24/2023, 9:40:26 AM   ERROR [ExceptionsHandler] No metadata for "User" was found.
> ```
> 
> This needs some thinking.
> 
> I've also added you contributors, so if you see how this can be fixed, you can push to this branch. Also reviewing your repository that you've sent.
> 
> PS. Let's keep option `synchronize: false,`, since we are using migrations.

Thank you for trusting. I will review it.

**Comment by vygandas** [comment]
I suppose the trickiest part is that all commands should work from the root directory, which means starting `nx run api:...`. Nobody should be using nest cli directly from `apps/api`, because that's kind of antipattern of the NX monorepo, where DTOs, interfaces and other things are shared across other apps. Just thought it's worth mentioning. All these commands are configured in `project.json` :)

**Comment by vygandas** [comment]
This might have good ideas https://medium.com/@arkadi.berikashvili/how-to-use-nestjs-cli-with-nx-workspace-40ebcfbea303

**Comment by llong2195** [comment]
> This might have good ideas https://medium.com/@arkadi.berikashvili/how-to-use-nestjs-cli-with-nx-workspace-40ebcfbea303

Correct. I am also using nx with nestjs cli

**Comment by llong2195** [comment]
@vygandas 
- Injecting a dataSource works well because 'autoLoadEntities: true' is a configuration of @nestjs/typeorm, whereas for dataSource, it doesn't, as TypeOrm doesn't have it.
- I think you should clearly define the names of entities and their types. When it's not clear, it can lead to various issues with querying, as seen in the error message:
`query failed: SELECT "User"."id" AS "User_id", "User"."firstName" AS "User_firstName", "User"."lastName" AS "User_lastName", "User"."email"
 AS "User_email", "User"."password" AS "User_password" FROM "user" "User"`
`error: relation "user" does not exist`

**Comment by llong2195** [comment]
@vygandas I have switched to using nestjs/cli. Please take a look and provide me with feedback. Thank you very much.

**Comment by vygandas** [comment]
@llong2195 nice, looking into it

**Comment by vygandas** [comment]
@llong2195 , I like the idea of introducing Entity naming conventions like UserEntity. Much clearer and consistent with other names, like UserInterface, UserCreateDto, etc. Also build seems to be creating separate `.js` files in `dist` as it should be. Also checked other changes and it all looks good.

Good catch with Entity naming, too. 👌🏻 👍🏻 

So now, what's left is to clean up and implement tests. Now tests for API seem to be failing, for example, here https://github.com/cortip/isomera/actions/runs/6987539802/job/19014184646?pr=93

Let me know if you are interested in looking at tests or I should take it.

**Comment by llong2195** [comment]
@vygandas Thanks for the positive feedback. I'm glad to have been able to contribute a bit. Could you help me with writing tests? I'm not very proficient in writing tests. Thank you so much

**Comment by vygandas** [comment]
@llong2195 I just commented out tests that were failing. Will create another ticket :) Thought this is blocking everything else, haha. Thanks for help.

**Pull request merged**